### PR TITLE
MM-14410: Adds the ability to add and manage groups for a team.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary

Adds the ability to add and manage groups for a team.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-14410

#### Related Pull Requests

- Has server changes (https://github.com/mattermost/mattermost-server/pull/10701)
- Has redux changes (https://github.com/mattermost/mattermost-redux/pull/821)

#### Screenshots

Screen Shot 2019-05-02 at 11 10 32 AM
Screen Shot 2019-05-02 at 11 10 48 AM
Screen Shot 2019-05-02 at 11 10 59 AM

### CODE REVIEW

File: components/add_groups_to_team_modal/add_groups_to_team_modal.jsx [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+                        buttonSubmitLoadingText={buttonSubmitLoadingText}
+                        saving={this.state.saving}
+                        loading={this.state.loadingGroups}
+                        placeholderText={localizeMessage('multiselect.addGroupsPlaceholder', 'Search and add groups')}
+                    />
+                </Modal.Body>
+            </Modal>
+        );
+    }
+}
```
Comment (mkraft) [inline components/add_groups_to_team_modal/add_groups_to_team_modal.jsx]: This component is a fairly straight copy of components/add_users_to_team/add_users_to_team.jsx.

File: components/add_groups_to_team_modal/add_groups_to_team_modal.jsx [hunk 2]
```diff
@@ -0,0 +1,278 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (mkraft) [inline components/add_groups_to_team_modal/add_groups_to_team_modal.jsx]: This component is a fairly straight copy of components/add_users_to_team/add_users_to_team.jsx.

File: components/add_groups_to_team_modal/index.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (mkraft) [inline components/add_groups_to_team_modal/index.js]: This is a fairly straight copy of components/add_users_to_team/index.js.

File: components/create_post/create_post.test.jsx [hunk 4]
```diff
@@ -27,7 +27,7 @@ jest.mock('react-dom', () => ({
     }),
 }));
 
-jest.mock('actions/post_actions.jsx', () => ({
+jest.mock('actions/post_actions.jsx', (process) => ({
```
Comment (mkraft) [inline components/create_post/create_post.test.jsx]: This existing code was causing the hoist checker to fail.

File: components/list_modal.jsx [hunk 5]
```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (mkraft) [inline components/list_modal.jsx]: This is a new reusable component that we can potentially use to replace all modals that look like this.

Screen Shot 2019-05-02 at 10 52 53 AM

File: components/team_groups_manage_modal/index.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (mkraft) [inline components/team_groups_manage_modal/index.js]: This is just a helper component to implement the `ListModal` component specifically for adding groups to a team.

File: components/add_groups_to_team_modal/add_groups_to_team_modal.jsx [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        }
+
+        let addError = null;
+        if (this.state.addError) {
+            addError = (<div className='has-error col-sm-12'><label className='control-label font-weight--normal'>{this.state.addError}</label></div>);
+        }
+
+        return (
+            <Modal
+                id='addUsersToTeamModal'
```
Comment (mkraft) [inline components/add_groups_to_team_modal/add_groups_to_team_modal.jsx]:
```suggestion
                id='addGroupsToTeamModal'
```

File: components/channel_selector_modal/channel_selector_modal.jsx [hunk 8]
```diff
@@ -222,6 +222,7 @@ export default class ChannelSelectorModal extends React.Component {
                         buttonSubmitText={buttonSubmitText}
                         saving={false}
                         loading={this.state.loadingChannels}
+                        placeholderText={localizeMessage('multiselect.placeholder', 'Search and add members')}
```
Comment (jespino) [inline components/channel_selector_modal/channel_selector_modal.jsx]: `Search and add members?` sounds confusing, it shouldn't be 'Search and add channels'

Comment (mkraft) [reply components/channel_selector_modal/channel_selector_modal.jsx]: This is the existing wording (as seen in this screenshot from the community server):

Screen Shot 2019-05-09 at 10 42 23 AM

File: components/create_post/create_post.test.jsx (hunk 4 above)
Comment (jespino) [reply components/create_post/create_post.test.jsx]: Yes, but I think your proposal is not correct, the jest.mock second parameter is a factory that receive no parameters and returns the mock, so... that way process should be undefined, not the "process" as expected.

Comment (mkraft) [inline components/create_post/create_post.test.jsx]:
```suggestion
jest.mock('actions/post_actions.jsx', () => ({
```

File: components/list_modal.jsx [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+        /**
+         * searchPlaceholderText (optional) a string to show as a placeholder in the search input.
+         */
+        searchPlaceholderText: PropTypes.string,
+
+        /**
+         * titleBarButton (optional) an object representing a title button.
+         */
+        titleBarButton: PropTypes.shape({
```
Comment (jespino) [inline components/list_modal.jsx]: You are passing here an object that you build each time when you use the component, and at the same time you are defining the component as PureComponent, so, or you improve how you pass this object to be sure the object is not changing, or you use the Component.

File: components/channel_selector_modal/channel_selector_modal.jsx (hunk 8 above)
Comment (jespino) [reply components/channel_selector_modal/channel_selector_modal.jsx]: I think the code here is not pointing to that modal, is the `components/channel_selector_modal/channel_selector_modal.jsx` with the title `Add Channels To **Channel Selection** List`

File: components/list_modal.jsx [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+                    />
+                </div>
+            );
+        }
+        return this.state.items.map((item) => (
+            this.props.renderRow(item, this)
+        ));
+    }
+
+    onNext = async () => {
```
Comment (jespino) [inline components/list_modal.jsx]: I don't see the point of the `async` here

File: components/list_modal.jsx [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        const nextPage = this.state.page + 1;
+        this.onPageChange(nextPage);
+    }
+
+    onPrev = async () => {
```
Comment (jespino) [inline components/list_modal.jsx]: and here

File: components/list_modal.jsx (hunk 9 above)
Comment (mkraft) [reply components/list_modal.jsx]: I agree with you in theory but in my testing the number of renders don't change whether it's the way it is now, whether it has a `shouldComponentUpdate` method as a `Component`, or whether it's a `PureComponent` with the `titleBarButtonText` and `titleBarButtonOnClick` params. I wonder if my test is broken.

File: components/list_modal.jsx (hunk 10 above)
Comment (mkraft) [reply components/list_modal.jsx]: It's not needed after I refactored this method, you're right.

File: components/list_modal.jsx (hunk 11 above)
Comment (mkraft) [reply components/list_modal.jsx]: same.

File: components/list_modal.jsx (hunk 9 above)
Comment (mkraft) [reply components/list_modal.jsx]: I wonder if the vDOM is handling it and avoiding the re-render but still causing extra cycles in the JS code.

File: components/channel_selector_modal/channel_selector_modal.jsx (hunk 8 above)
Comment (mkraft) [reply components/channel_selector_modal/channel_selector_modal.jsx]: You're right it's not that modal. It is the existing text for the add channels and teams to a group in system console but probably is wrong there. I'll update it.


Screen Shot 2019-05-09 at 12 01 01 PM

**Pull request merged**